crates/lsp/src/header_reader.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    io, mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncBufRead, Future};
+
+const HEADER_DELIMITER: &'static [u8; 4] = b"\r\n\r\n";
+
+pub fn read_headers<'a, R>(
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+) -> impl Future<Output = io::Result<usize>> + 'a
+where
+    R: AsyncBufRead + ?Sized + Unpin,
+{
+    HeaderReader {
+        reader,
+        buf,
+        read: 0,
+    }
+}
+
+struct HeaderReader<'a, R: ?Sized> {
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+    read: usize,
+}
+
+impl<'a, R: ?Sized> HeaderReader<'a, R> {
+    fn read_headers<C: AsyncBufRead + ?Sized>(
+        mut reader: Pin<&mut C>,
+        cx: &mut Context<'_>,
+        buf: &mut Vec<u8>,
+        read: &mut usize,
+    ) -> Poll<io::Result<usize>> {
+        loop {
+            let (done, used) = {
+                let available = futures::ready!(reader.as_mut().poll_fill_buf(cx))?;
+                if let Some(i) = find_delimeter(&available) {
+                    buf.extend_from_slice(&available[..i + 5]);
+                    (true, i + 5)
```

Comment:
Takes care of the happy path, where the delimiter is found in the available bytes.